### ISSUE TITLE
Overlay sometimes adds a lot of edges that will never appear in results, causes slowdown

### ISSUE THREAD

**Issue by amykglen** [RTXteam/RTX#1069]
due to the way the Overlay actions that add virtual edges are designed, I think they can end up consuming a lot of time/space creating virtual edges that will never actually appear in a result. example:

```
add_qnode(name=MONDO:0005136, id=n00)
add_qnode(type=protein, is_set=true, id=n01)
add_qnode(type=chemical_substance, id=n02)
add_qedge(source_id=n00, target_id=n01, id=e00)
add_qedge(source_id=n01, target_id=n02, id=e01)
expand(edge_id=[e00,e01], kp=ARAX/KG2)
overlay(action=compute_ngd, virtual_relation_label=N3, source_qnode_id=n01, target_qnode_id=n02)
resultify(ignore_edge_direction=true)
filter_results(action=limit_number_of_results, max_results=100)
```

the `overlay` call in this query will result in an O(n2) kind of situation (or, specifically: `number of n01 nodes in the KG` * `number of n02 nodes in the KG`), because of this line in `compute_ngd`:

https://github.com/RTXteam/RTX/blob/1c68279a867bd3b3462983cfc24adfd8f2c5c6bd/code/ARAX/ARAXQuery/Overlay/compute_ngd.py#L95

which considers all combinations of all source (n01) nodes and all target (n02) nodes in the KG, rather than only those node pairs that could actually possibly appear together in a result (i.e., are connected by a path matching that in the query graph).

KG1 queries are small enough that this generally doesn't seem to be a problem with those, but with KG2's size, it can become very slow (we're talking hours)... for example, if the above query is run using KG2, that results in 303 n01 nodes and 15392 n02 nodes, so 303 * 15392 = 4.7 million node pairs to compute NGD/create virtual edges between. but a small fraction of those virtual edges would actually appear in a `result`.

whereas if the pool of node pairs was narrowed down to only those connected in the KG (in a manner satisfying the QG) this number should be much lower (should be O(n), I think: the number of n02 nodes * the average number of n01 nodes per result - so something on the order of 50,000, perhaps).

I have an idea about how this could be done (narrowing down the node pairs, that is) - but wonder if others think this seems worthwhile to address? (most of the time it seems Overlay actions are used between an 'n00' node (i.e., a qnode for which there is only 1 node in the KG) and another node, so there isn't an issue there. (e.g., 1 * 15,392 = 15,392.) but it seems there is a use case for overlaying between 'n01' and 'n02' kinds of nodes.)

**Comment by edeutsch** [comment]
I heartily agree, something needs to be done. I independently tried switching the demo query from KG1 to KG2 and likewise my query seems to have died after 4 hours. or maybe something is still happening? Unclear. But I think it's not workable. Maybe this is what's killing the server.

I don't know what the solution is, but I agree it's a problem.

Note that this a key case where KG1 is still needed because attempts to use KG2 for this end in failure. We *want* to switch to KG2 for this because it has so much more information, but somehow we need to come up with clever ways to reduce the problem to a more manageable size.

**Comment by amykglen** [comment]
pushed a fix for this to `compute_ngd` in the `issue1069` branch - it uses `resultify` to identify which source/target nodes could appear in the same result, and then only computes NGD between those node pairs.

so, for the example I wrote up in this issue, instead of overlaying 4.7 million virtual edges (hours), it overlays 170,000 (takes 15 minutes when skipping eUtils calls for the 1% that are fastNGD 'misses'). yet NGD values in the results are the same as before.

waiting to merge this into `master` since we're very close to the start of a relay meeting, but I've tested locally and all seems well so far.

**Comment by dkoslicki** [comment]
I concur that this needs to be addressed @amykglen and glad you were able to find a fix for it. Indeed, sometimes (especially with our new ranker), calling `overlay` on other than the one start node is a good thing to do, and paring down the O(n^2) as much as possible is a smart thing to do. After tested, I definitely think it should be merged to master

**Comment by amykglen** [comment]
ok, this has been merged into master (for `compute_ngd` - not implemented in other `overlay` modules).

**Comment by amykglen** [comment]
ok, so the solution indeed seems to be working well on test/production in `compute_ngd`.

it'd probably make sense to switch the other overlay modules to use this as well at some point - I went ahead and did so for `overlay_exposures_data`, and in doing so, realized it would make sense to have this function in a common place, so that the various overlay classes (`ComputeNGD`, `OverlayExposuresData`, etc.) can all use it. so I created [overlay_utilities.py](https://github.com/RTXteam/RTX/blob/master/code/ARAX/ARAXQuery/Overlay/overlay_utilities.py) in the `ARAXQuery/Overlay/` directory to host it.

it's now fairly easy to plug into an overlay module (see [this commit](https://github.com/RTXteam/RTX/commit/1cbec77ecb6423d5d4f637561416b81b841a6bcb) as an example). it's basically just a matter of importing `overlay_utilities.py` and then swapping out the classic line:
```
for source_curie, target_curie in itertools.product(source_curies, target_curies):
```
with:
```
for source_curie, target_curie in ou.get_node_pairs_to_overlay(source_qnode_id, target_qnode_id, query_graph, knowledge_graph, self.response):
```
changes are in `master` (tested locally).
